File: UnrealClient/Source/UnrealClient/java_to_cpp_model/model/GameInstance.cpp
//====================================================================================================
//The Free Edition of Java to C++ Converter limits conversion output to 100 lines per file.

//To purchase the Premium Edition, visit our website:
//https://www.tangiblesoftwaresolutions.com/order/order-java-to-cplus.html
//====================================================================================================

#include "GameInstance.h"
#include "TemplePlayerData.h"

namespace com::example::model
{
	using namespace com::fasterxml::jackson::annotation;

	GameInstance::GameInstance()
	{
	}

	GameInstance::GameInstance(GameState gameState, std::optional<int> &ID, std::unordered_map<int, std::shared_ptr<PlayerData>> &players)
	{
		this->gameState = gameState;
		this->gameID = ID.value();

		if (players.empty())
		{
			this->players = std::unordered_map<int, PlayerData>();
		}
		else
		{
			this->players = players;
			this->gameOwner = (std::vector<>(players.values()))[0];
		}
	}

	GameInstance::GameInstance(GameState gameState, std::optional<int> &ID, std::unordered_map<int, std::shared_ptr<PlayerData>> &players, std::shared_ptr<PlayerData> gameOwner)
	{
		this->gameState = gameState;
		this->gameID = ID.value();

		if (gameOwner == nullptr)
		{
			throw std::runtime_error("players map cannot be null");
		}
		// concrete class in GameInstance?
		this->gameOwner = std::static_pointer_cast<TemplePlayerData>(gameOwner);

		// set game name
		if ((std::static_pointer_cast<TemplePlayerData>(gameOwner))->gameName != L"")
		{
			this->gameName = (std::static_pointer_cast<TemplePlayerData>(gameOwner))->gameName;
		}
		else
		{
			this->gameName = L"TempleRun-" + ID.value();
		}

		// create new players map or set players map
		if (players.empty())
		{
			this->players = std::unordered_map<int, PlayerData>();
		}
		else
		{
			this->players = players;
		}
		this->players.emplace(gameOwner->getPublicID(), gameOwner);

	}

	std::wstring GameInstance::getGameName()
	{
		return gameName;
	}

	std::shared_ptr<GameInstance> GameInstance::setGameName(const std::wstring &gameName)
	{
		this->gameName = gameName;
		return shared_from_this();
	}

	std::shared_ptr<PlayerData> GameInstance::getGameOwner()
	{
		return gameOwner;
	}

	void GameInstance::setGameOwner(std::shared_ptr<PlayerData> gameOwner)
	{
		this->gameOwner = gameOwner;
	}

	bool GameInstance::addPlayer(std::shared_ptr<PlayerData> player)
	{
		if (player == nullptr)
		{
			throw std::runtime_error("Null Input Exception");
		}
		if (std::dynamic_pointer_cast<TemplePlayerData>(player) != nullptr)
		{
			std::shared_ptr<TemplePlayerData> other = std::static_pointer_cast<TemplePlayerData>(player);

			if (players.empty())
			{
				setGameOwner(player);
			}
			if (players.find(other->getPublicID()) != players.end())
			{
				return false;
			}

			return this->players.emplace(other->getPublicID(), other) == nullptr;
		}
		return false;
	}

	bool GameInstance::updatePlayer(std::shared_ptr<PlayerData> player)
	{
		if (player == nullptr)
		{
			throw std::runtime_error("Null Input Exception");
		}
		if (players.find(player->getPublicID()) == players.end())
		{
			return false;
		}

		return this->players.replace(player->getPublicID(), player) != nullptr;
	}

	bool GameInstance::removePlayer(int playerPublicID)
	{
		if (this->players.find(playerPublicID) == this->players.end())
		{
			return false;
		}
		else
		{
			return this->players.erase(playerPublicID) != nullptr;
		}
	}

//JAVA TO C++ CONVERTER TODO TASK: Most Java annotations will not have direct C++ equivalents:
//ORIGINAL LINE: @JsonIgnore public boolean isEmpty()
	bool GameInstance::isEmpty()
	{
		return this->players.empty();
	}

//JAVA TO C++ CONVERTER TODO TASK: Most Java annotations will not have direct C++ equivalents:
//ORIGINAL LINE: @JsonIgnore public boolean isFull()
	bool GameInstance::isFull()
	{
		return this->players.size() >= 4;
	}

	bool GameInstance::join()
	{
		this->setGameState(GameState::GAME_LOBBY);
		return true;
	}

	bool GameInstance::start()
	{
		this->setGameState(GameState::GAME_STARTING);
		return true;
	}

	bool GameInstance::enter()
	{
		this->setGameState(GameState::ENTERING_GAME);
		return true;
	}

	bool GameInstance::run()
	{
		this->setGameState(GameState::IN_GAME);
		return true;
	}

	bool GameInstance::close()
	{
		this->setGameState(GameState::CLOSING);
		return true;
	}

	int GameInstance::getGameID()
	{
		return this->gameID;
	}

	void GameInstance::setGameID(int iD)
	{
		this->gameID = iD;
	}

	GameInstance::GameState GameInstance::getGameState()
	{
		return this->gameState;
	}

	void GameInstance::setGameState(GameState newState)
	{
		this->gameState = newState;
	}

	std::shared_ptr<PlayerData> GameInstance::getPlayer(int playerID)
	{
		return std::static_pointer_cast<TemplePlayerData>(this->players[playerID]);
	}

	std::unordered_map<int, std::shared_ptr<PlayerData>> GameInstance::getPlayers()
	{
		return players;
	}

	std::wstring GameInstance::toString()
	{
		return L"gameID=" + std::to_wstring(gameID) + L", gameName=" + gameName + L", gameOwner=" + gameOwner->getName() + L", gameState=" + gameState;

	}

	bool GameInstance::equals(std::any o)
	{
		if (shared_from_this() == o)
		{
			return true;
		}
		if (!(o.type() == typeid(std::shared_ptr<GameInstance>)))
		{
			return false;
		}
		std::shared_ptr<GameInstance> game = std::any_cast<std::shared_ptr<GameInstance>>(o);
		if (gameID != game->gameID)
		{
			return false;
		}
		if (gameState != game->gameState)
		{
			return false;
		}

//====================================================================================================
//End of the allowed output for the Free Edition of Java to C++ Converter.

//To purchase the Premium Edition, visit our website:
//https://www.tangiblesoftwaresolutions.com/order/order-java-to-cplus.html
//====================================================================================================

File: server/model/src/main/java/com/example/model/GameInstance.java
package com.example.model;

import com.fasterxml.jackson.annotation.*;


import java.util.*;
import java.util.stream.*;


public class GameInstance {

    private int gameID;

    private GameState gameState;

    public String gameName;

    @JsonIgnore
    private PlayerData gameOwner;

    private Map<Integer, PlayerData> players;

    public GameInstance(){}
    
    public GameInstance(GameState gameState, Integer ID, Map<Integer, PlayerData> players) throws Exception {
        this.gameState = gameState;
        this.gameID = ID;

        if(players == null) {
            this.players = new HashMap<>();
        } else {
            this.players = players;
            this.gameOwner = new ArrayList<>(players.values()).get(0);
        }
    }

    public GameInstance(GameState gameState, Integer ID, Map<Integer, PlayerData> players, PlayerData gameOwner) throws Exception {
        this.gameState = gameState;
        this.gameID = ID;

        if(gameOwner == null) {
            throw new Exception("players map cannot be null");
        }
        // concrete class in GameInstance?
        this.gameOwner = (TemplePlayerData) gameOwner;

        // set game name
        if(((TemplePlayerData) gameOwner).gameName != null) {
            this.gameName = ((TemplePlayerData) gameOwner).gameName;
        }
        else {
            this.gameName = "TempleRun-" + ID;
        }

        // create new players map or set players map
        if(players == null) {
            this.players = new HashMap<>();
        }
        else {
            this.players = players;
        }
        this.players.put(gameOwner.getPublicID(), gameOwner);

    }
    
    public String getGameName() {
        return gameName;
    }

    public GameInstance setGameName(String gameName) {
        this.gameName = gameName;
        return this;
    }

    public PlayerData getGameOwner() {
        return gameOwner;
    }

    private void setGameOwner(PlayerData gameOwner) throws Exception {
        this.gameOwner = gameOwner;
    }

    public boolean addPlayer(PlayerData player) throws Exception {
        if(player == null){
            throw new Exception("Null Input Exception");
        }
        if(player instanceof TemplePlayerData) {
            TemplePlayerData other = (TemplePlayerData)  player;

            if(players.isEmpty()){
                setGameOwner(player);
            }
            if (players.containsKey(other.getPublicID())) {
                return false;
            }

            return this.players.put(other.getPublicID(), other) == null;
        }
        return false;
    }

    public boolean updatePlayer(PlayerData player) throws Exception {
        if(player == null){
            throw new Exception("Null Input Exception");
        }
        if(!players.containsKey(player.getPublicID())){
            return false;
        }

        return this.players.replace(player.getPublicID(), player) != null;
    }

    public boolean removePlayer(int playerPublicID) {
        if (!this.players.containsKey(playerPublicID)) {
            return false;
        } else {
            return this.players.remove(playerPublicID) != null;
        }
    }
    
    @JsonIgnore
    public boolean isEmpty() {
        return this.players.isEmpty();
    }
    @JsonIgnore
    public boolean isFull() {
        return this.players.size() >= 4;
    }

    public boolean join() {
        this.setGameState(GameState.GAME_LOBBY);
        return true;
    }

    public boolean start() {
        this.setGameState(GameState.GAME_STARTING);
        return true;
    }

    public boolean enter() {
        this.setGameState(GameState.ENTERING_GAME);
        return true;
    }

    public boolean run() {
        this.setGameState(GameState.IN_GAME);
        return true;
    }

    public boolean close() {
        this.setGameState(GameState.CLOSING);
        return true;
    }


    public int getGameID() {
        return this.gameID;
    }

    public void setGameID(int iD) {
        this.gameID = iD;
    }

    public GameState getGameState() {
        return this.gameState;
    }

    public void setGameState(GameState newState) {
        this.gameState = newState;
    }

    public PlayerData getPlayer(int playerID) {
        return (TemplePlayerData)this.players.get(playerID);
    }

    public Map<Integer, PlayerData> getPlayers() {

        return players;
    }

    @Override
    public String toString() {
        return
                "gameID=" + gameID +
                        ", gameName=" + gameName +
                        ", gameOwner=" + gameOwner.getName() +
                        ", gameState=" + gameState;
    
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof GameInstance)) return false;
        GameInstance game = (GameInstance) o;
        if (gameID != game.gameID) return false;
        if (gameState != game.gameState) return false;
        return Objects.equals(players, game.players);
    }

    @Override
    public int hashCode() {
        int result = gameState != null ? gameState.hashCode() : 0;
        result = 31 * result + gameID;
        result = 31 * result + (players != null ? players.hashCode() : 0);
        return result;
    }

    private void setCustomGameName(String gameName) {
        this.gameName = gameName;
    }

    public static enum GameState {
        GAME_LOBBY("GAME_LOBBY"),
        GAME_STARTING("GAME_STARTING"),
        ENTERING_GAME("ENTERING_GAME"),
        IN_GAME("IN_GAME"),

        GAME_ENDING("GAME_ENDING"),
        CLOSING("CLOSING"),

        UNKNOWN("GAME_STARTING");
        public final String state;
        private GameState(String state) {
            this.state = state;
        }
    }
}

